### ISSUE TITLE
ref as a function breaks resizing

### ISSUE THREAD

**Issue by simonsmith** [rpearce/react-expanding-textarea#52]
Hello,

If I pass a function as the `ref` the resizing breaks, and I presume this is because the textarea relies on the `ref.current` to be an element. However libraries like `react-hook-form` pass a `register` function as the ref which gets the element as the only argument.

Thanks

**Comment by rpearce** [comment]
@simonsmith I'm not sure at the moment if I'll want to support the old ref-as-a-function technique.

As this library has a React peer dependency of >=16.8, you should have access to `createRef` (just as this lib uses [here](https://github.com/rpearce/react-expanding-textarea/blob/ce686cc4dff017ea5910c05a2575e9529edb2420/source/index.tsx#L74)), and that is an object with `current` inside of it, just like `useRef` gives you.

Example off the top of my head if you're using a `class`:
```js
import React, { Component, createRef } from 'react'
import Textarea from 'react-expanding-textarea'

class MyComp extends Component {
  constructor(...params) {
    super(...params)
    this.myRef = createRef()
  }

  render() {
    return (
      <Textarea ref={this.myRef} ... />
    )
  }
}
```

**Comment by simonsmith** [comment]
That's no good for me unfortunately but I've found a workaround with `react-hook-form`. I'll post it here as reference for anyone else:

```tsx
// This is what breaks currently, but `register` is needed for react-hook-form
<ExpandingTextarea ref={register({required: true})} name="some-field" />;

// Instead it can be tracked manually

const {register, setValue} = useForm();

React.useEffect(() => {
  register({name: 'some-field'}, {required: true});
}, []);

<ExpandingTextarea
  onChange={(event: React.FormEvent<HTMLTextAreaElement>) => {
    setValue('some-field', event.currentTarget.value);
  }}
/>;
```

Cheers

**Comment by jordie23** [comment]
Thanks for posting your work around @simonsmith I just ran into this exact same issue, using react-hook-form.

@rpearce Firstly, thanks for this package, it's great! If you have free time I would suggest looking into support for it as react-hook-form is quickly becoming widely adopted. You can see the trend graph here surging https://www.npmjs.com/package/react-hook-form - it's at 500,000+ downloads per week atm.

**Comment by rpearce** [comment]
Ok, I've reopened this and will think more about how to do this.

@jordie23 can you please provide an example, using `react-hook-form`, of how you're using these two together? This will allow me to easily get going and test out solutions.

**Comment by jordie23** [comment]
Thanks @rpearce 

I've come up with an example here. https://codesandbox.io/s/react-hook-form-basic-forked-qb778?file=/src/index.js

It shows how react-hook-form is used with a normal input and textareas. You'll see the field registered with the default `ref={register}` and with validation in the form of `ref={register({required: true})}`. I've then included your expanding text box, once without the registration to show it working, then another 2 times with the two ^ reg= variations above, showing how the expansion no long works. Click submit to see the output (not broken).

**Comment by rpearce** [comment]
@jordie23 With my 8 week old child needing attending to every little while, I haven't found time to work on this yet. It's on my mind, though.

**Comment by rpearce** [comment]
I'm looking at this and am not yet sure how to forward a function ref _that I can also get access to_.

https://github.com/react-hook-form/react-hook-form/blob/8c198d88cb46e55a7ef709767726e6e8c56a3eab/src/useForm.ts#L1283

👆 This code from `react-hook-form` shows that it is indeed a callback function and also leads down a bit of a rabbit hole for trying to figure out what it does.

Ultimately, I need to figure out:
a. how to access a function ref's DOM element value (if that's even possible?)
b. if I can't access it, I need to see about another way around this (getting hacky with `querySelector`?? TBD)

**Comment by jordie23** [comment]
@rpearce There’s no rush on this. Congrats on the baby :) we have a work around so don’t put too much time in if you don’t have any right now.

react-hook-form has another way of working, using its Controller component I’ve found out. I wonder if that’s just what we need to use here. (As in, you probably don’t need to do anything and we need to use this method) https://react-hook-form.com/api/#Controller

**Comment by jordie23** [comment]
I’ll have a play around and see if I can come up with something. If I can I’ll post here and maybe the solution here is to add the code sample to the docs

**Comment by rpearce** [comment]
Thanks! Okay, that works for me.

Alternatively, here's the 1 concept that would need to get worked around to make this work:

https://github.com/rpearce/react-expanding-textarea/blob/main/source/index.tsx#L79

That `ref.current` is how this lib accesses the textarea DOM node, so if there's a way to access that DOM node _and_ allow for whatever ref you pass to get applied, then 🎉 .

**Comment by rpearce** [comment]
In other news, I've added @simonsmith and @jordie23 to the all contributors list as of https://github.com/rpearce/react-expanding-textarea/pull/54/files.

**Comment by simonsmith** [comment]
> That ref.current is how this lib accesses the textarea DOM node, so if there's a way to access that DOM node and allow for whatever ref you pass to get applied, then

I haven't tested this, but could you do something like:

```js
const Input = React.forwardRef((props, forwardedRef) => {
  const ref = React.useRef(forwardedRef.current);
  if (typeof forwardedRef === 'function') {
    forwardedRef(ref);
  }
  return <input ref={ref} type="text" />;
});
```

Always create a `ref` and set its default value if the user passes one in from the outside. If it's a function (for RHF) then pass it along.

**Comment by rpearce** [comment]
Alright, I'll take a look. Thank you for the help

**Comment by rpearce** [comment]
@simonsmith @jordie23 would y'all mind checking out https://github.com/rpearce/react-expanding-textarea/pull/55 and seeing if the test and new storybook story make sense and if this concept will work for you?

**Comment by rpearce** [comment]
This should be closed by #55 and released in version `2.3.0`!

Please let me know if there are any further issues. Thanks for the help!

**Comment by jordie23** [comment]
@rpearce Sorry for the delayed reply. This does fix it for me. Thanks! Appreciate the effort a lot.
